Refactor auth gRPC middleware and transport credentials

rosstimothy opened the pull request:
Adds a new `credentials.TransportCredentials` to auth and refactors
`auth.Middleware` to make use of them. This allows for less redundancy
per RPC and provides the ability to perform session control when grpc
connections are established.

The `auth.Middleware.withAuthenticatedUser` function was called from all
interceptors of any gRPC server that leveraged `auth.Middleware`. This
was constructing the identity of the client from the TLS certificate
per RPC. Since the identity of a connection cannot change between RPCs
there is no need to construct the identity in an interceptor. Instead,
the new `auth.TransportCredentials` can build the identity once after the
TLS handshake completes. When interceptors or gRPC handlers need to
acquire the identity they can do so using the same
mechanism via `peer.Peer.AuthInfo` and assert on the new `auth.IdentityInfo`
type:

```go
    peerInfo, ok := peer.FromContext(ctx)
	if !ok {
		return
	}

	connInfo, ok := peerInfo.AuthInfo.(auth.IdentityInfo)
	if !ok {
	    return
	}

	identity := connInfo.IdentityGetter.GetIdentity()
	connState := connInfo.TLSInfo.State
```

The gRPC server that is served on the Proxy ssh port needs to hold the
same invariants as the ssh server available on the same port. This
means that connection limits need to be enforced when new connections
are established. To achieve this the new `auth.TransportCredentials` can
be provided with an `auth.ConnectionEnforcer` and an `auth.Authorizer`
to first build an identity, then ensures that the user is authorized and
finally that the user has not exceeded their connection limits.

Other minor changes include providing the cluster name to the
`auth.Middleware` instead of it looking the value up from the cache every
time an identity is created. Some interceptors were also moved around or
removed to eliminate the chances of having duplicate interceptors end up in
the interceptor chain.

Part of #19812

smallinsky commented on lib/auth/transport_credentials.go:
```diff
@@ -0,0 +1,234 @@
+// ConnectionIdentity contains the identifying properties of a
+// client connection required to enforce connection limits.
+type ConnectionIdentity struct {
+	Username      string
+	AccessChecker services.AccessChecker
+	LocalAddr     string
+	RemoteAddr    string
+	UserMetadata  apievents.UserMetadata
```
not: godocs.

smallinsky commented on lib/auth/transport_credentials.go:
```diff
@@ -0,0 +1,243 @@
+
+	identityGetter, err := c.userGetter.GetUser(tlsInfo.State)
+	if err != nil {
+		conn.Close()
```
nit: to prevent placing conn.Close call in each error != case we can just 
add a helper function like:
```go
(c *TransportCredentials) ServerHandshake(rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
   conn, authInfo, err := c.serverHandshake(rawConn) 
   if err != nil {
     conn.Close()
     return nil, nil. trace.Wrap(err)
   }
}
```

smallinsky reviewed (approved):
I have left one additional [comment regarding returning nil,nil](https://github.com/gravitational/teleport/pull/21850/files#diff-85f8807d6d29016f5842ed44f60267ad40e8438ff656b83274f2b0f13e2df536R199) that I would like to clarify before merging.  

Apart aforementioned issue LGTM.
Thanks for addressing my nit-picking comments.  I think that moving the auth part to `ServerHandshake` and preforming auth logic only once per connection is clever idea.

smallinsky commented:
@rosstimothy 
> @smallinsky I think I've addressed all your concerns. I'll leave this up even though you already approved to make sure that you are satisfied with the last round of changes before merging

I don't have any additional comments and current version LGTM

rosstimothy committed:
Adds a new `credentials.TransportCredentials` to auth and refactors
`auth.Middleware` to make use of them. This allows for less redundancy
per RPC and provides the ability to perform session control when grpc
connections are established.

The `auth.Middleware.withAuthenticatedUser` function was called from all
interceptors of any gRPC server that leveraged `auth.Middleware`. This
was constructing the identity of the client from the TLS certificate
per RPC. Since the identity of a connection cannot change between RPCs
there is no need to construct the identity in an interceptor. Instead,
the new `auth.TransportCredentials` can build the identity once after the
TLS handshake completes. When interceptors or gRPC handlers need to
acquire the identity they can do so using the same
mechanism via `peer.Peer.AuthInfo` and assert on the new `auth.IdentityInfo`
type:

```go
    peerInfo, ok := peer.FromContext(ctx)
	if !ok {
		return
	}

	connInfo, ok := peerInfo.AuthInfo.(auth.IdentityInfo)
	if !ok {
	    return
	}

	identity := connInfo.IdentityGetter.GetIdentity()
	connState := connInfo.TLSInfo.State
```

The gRPC server that is served on the Proxy ssh port needs to hold the
same invariants as the ssh server available on the same port. This
means that connection limits need to be enforced when new connections
are established. To achieve this the new `auth.TransportCredentials` can
be provided with an `auth.ConnectionEnforcer` and an `auth.Authorizer`
to first build an identity, then ensures that the user is authorized and
finally that the user has not exceeded their connection limits.

Other minor changes include providing the cluster name to the
`auth.Middleware` instead of it looking the value up from the cache every
time an identity is created. Some interceptors were also moved around or
removed to eliminate the chances of having duplicate interceptors end up in
the interceptor chain.

Part of #19812
